[libc][mathvec] Initial commit for LIBC vector math component

DylanFleming-arm opened the pull request:
Created mathvec directories and unittest framework for vector math functions, as well as an initial implementation of vector expf, which is presently CR for round-to-nearest.

blapie committed:
Add simple vector math tester

Simplistic checks based on a wrapper duplicating a scalar input.

Introduce SIMD matcher (ultimately should be moved to test utils/)
- Comparison with exception is supported.
- Test all rounding modes in one line.
- Control lanes

A potential improvement could be made later: load vector in test
over range instead of duplicating input. That would reduce run
time of vector testing.

DylanFleming-arm committed:
Initial commit for mathvec directories

Created a mathvec/ file structure that mirrors the
structure of the math/ directory, and added a dummy
expf implementation.

DylanFleming-arm committed:
Add CR (RN only) vector expf

Implements a fully vectorised expf routine, that is
correctly rounded for Round-to-nearest. As an example
for what a correctly rounded vector routine looks like
using LLVM LIBC SIMD types.

jhuber6 commented on libc/src/CMakeLists.txt:
```diff
@@ -29,6 +29,11 @@ if(${LIBC_TARGET_OS} STREQUAL "linux")
 endif()
 
+if(LIBC_COMPILER_HAS_EXT_VECTOR_TYPE)
+  message(STATUS "Vector math enabled")
```
This should either be removed for made to be more descriptive, don't think it means much if there's no way to turn it off.

DylanFleming-arm replied:
Oops, this was more for debugging while getting stuff set up. I've removed it.

jhuber6 commented on libc/src/__support/mathvec/vector_utils.h:
```diff
@@ -0,0 +1,46 @@
+
+// Casts a simd<float, N> into two simd<double, N/2>
+template <size_t N>
+LIBC_INLINE constexpr auto vector_float_to_double(cpp::simd<float, N> v) {
```
I'm pretty sure the compiler should correctly handle oversized vectors, would be easier to just do the regular SIMD cast I would think.

DylanFleming-arm replied:
Ah you're right, makes things a lot clearer.

SchrodingerZhu commented on libc/src/__support/mathvec/expf_utils.h:
```diff
@@ -0,0 +1,29 @@
+      true, index, common_constants_internal::EXP_MANTISSA);
+  auto exponent = (u >> 6) << 52;
+  auto result = mantissa | exponent;
+  return reinterpret_cast<cpp::simd<double, N>>(result);
```
It feels strange to do `reinterpret_cast` from a simd register of integer to a simd register of double.

```suggestion
  return cpp::bit_cast<cpp::simd<double, N>>(result);
```

jhuber6 replied:
Yeah, bit casting here seems suspicious. I'd assume it'd either be a broadcast or a simd cast

DylanFleming-arm replied:
Ah, arm C intrinsics call this a reinterpret cast, I wasn't aware of the C++ distinction. Thanks !

DylanFleming-arm committed:
Removed vector_utils, Improved Gathers, and fixed bitcast.

Removes vector_utils.h, and instead relies on simd_cast/compiler
to correctly handle oversized vectors.

Added check for gather instruction support, and implemented
an explicit loop over scalar lookup fallback.

Changed incorrect reinterpret_cast to bit_cast.

Removed leftover debug message from a CMakeList file.

DylanFleming-arm commented on libc/src/__support/mathvec/common_constants.h:
```diff
@@ -0,0 +1,40 @@
+
+// Lookup table for mantissas of 2^(i / 64) with i = 0, ..., 63.
+static constexpr uint64_t EXP_MANTISSA[64] = {
+    0x0000000000000, 0x02c9a3e778061, 0x059b0d3158574, 0x0874518759bc8,
```
While they represent floats, they're used in bit manipulation directly after their lookup, so I have them store as integers to remove the extra bitcast. I'm happy to make these hexfloat doubles if it's more the more expected approach.

DylanFleming-arm commented on libc/src/__support/mathvec/expf_utils.h:
```diff
@@ -0,0 +1,29 @@
+template <size_t N>
+LIBC_INLINE cpp::simd<double, N> exp_lookup(cpp::simd<uint64_t, N> u) {
+  auto index = u & cpp::simd<uint64_t, N>(0x3f);
+  auto mantissa = cpp::gather<cpp::simd<uint64_t, N>>(
```
Good point, I added a fallback into the simd.h gather for targets that don't support it.

jhuber6 commented on libc/src/__support/mathvec/expf.h:
```diff
@@ -0,0 +1,69 @@
+  auto p04 = p01 + r2 * p23;
+  auto y = r + p04 * r2;
+
+  auto u = reinterpret_cast<cpp::simd<uint64_t, N>>(z);
```
Is this rounding the floating point value to an integer for a LUT? It'd probably be better to use one of the elementwise floor / trunc operations then just assign it to an integer vector of the same size.

jhuber6 commented on libc/src/__support/mathvec/expf.h:
```diff
@@ -0,0 +1,69 @@
+  cpp::simd<double, N> x_d = cpp::simd_cast<double, float, N>(x);
+  cpp::simd<double, N> y = inline_exp(x_d);
+  cpp::simd<float, N> ret = cpp::simd_cast<float, double, N>(y);
+  return cpp::select(is_special, special_res, ret);
```
```suggestion
  return is_special ? special_res : ret;
```
Let's just use the ternary, I should probably just remove `select`. The `simd.h` from C++ uses `where`, but we don't support that because it requires a random object.

jhuber6 commented on libc/src/mathvec/expf.h:
```diff
@@ -0,0 +1,21 @@
+
+namespace LIBC_NAMESPACE_DECL {
+
+cpp::simd<float> expf(cpp::simd<float> x);
```
Does anyone know how the mangling for these will work? I'm assuming we'd want to build multiple variants for each different ABI into a single library?

DylanFleming-arm committed:
Updated coefficients and addressed comments

Updated coefficients used in expf, and added
sollya comments to reproduce them.

Made inline_exp and exp_lookup static to fix
ABI warning.

Reduced number of auto types used.

jhuber6 reviewed (approved):
This looks good to me usage-wise, I'll leave the math to @lntue

DylanFleming-arm committed:
Made cpp::bit_cast static

michaelrj-google commented on libc/test/src/mathvec/AddTest.h:
```diff
@@ -0,0 +1,87 @@
+//===-- Utility class to test different flavors of float add ----*- C++ -*-===//
```
is this file used anywhere? I don't see it in the cmake.

jhuber6 commented on libc/lib/CMakeLists.txt:
```diff
@@ -2,10 +2,10 @@ set(libc_archive_targets "")
+  list(APPEND libc_archive_targets libc libm libmvec)
   list(APPEND libc_archive_entrypoint_lists
-       TARGET_LIBC_ENTRYPOINTS TARGET_LIBM_ENTRYPOINTS)
+       TARGET_LIBC_ENTRYPOINTS TARGET_LIBM_ENTRYPOINTS TARGET_LIBMVEC_ENTRYPOINTS)
```
The standard is to put this in a separate library that people can select with https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fveclib but we could equivalently just put everything in one library and have that option just link the LLVM libm.

lntue commented on libc/src/mathvec/generic/expf.cpp:
```diff
@@ -0,0 +1,18 @@
+
+namespace LIBC_NAMESPACE_DECL {
+
+LLVM_LIBC_FUNCTION(cpp::simd<float>, expf, (cpp::simd<float> x)) {
```
Do we want to create an `expf` C symbol for this function in the public package `libmvec.a`?  If not then `LLVM_LIBC_FUNCTION` macro should be removed, and just use normal function declaration.

DylanFleming-arm committed:
Minor structure/build system changes

Moved mathvec/common_constants.h and mathvec/expf_utils.h
into the mathvec namespace, removed redundant file, and
set TARGET_LIBMVEC_ENTRYPOINTS to be created empty when
the target doesn't support vector types.

DylanFleming-arm committed:
Removed LLVM_LIBC_FUNCTION macro usage

jhuber6 commented on libc/src/mathvec/generic/expf.cpp:
```diff
@@ -11,7 +11,7 @@
 namespace LIBC_NAMESPACE_DECL {
 
-LLVM_LIBC_FUNCTION(cpp::simd<float>, expf, (cpp::simd<float> x)) {
+cpp::simd<float> expf (cpp::simd<float> x) {
```
We'll probably  want a special version for these kinds of variants. Once this PR lands I can look into it since we probably want something similar for GPU math as well.

DylanFleming-arm committed:
Fixed clang format issues

bassiounix commented on libc/src/__support/mathvec/expf.h:
```diff
@@ -0,0 +1,84 @@
+  cpp::simd<uint64_t, N> u = cpp::bit_cast<cpp::simd<uint64_t, N>>(z);
+  cpp::simd<double, N> s = exp_lookup(u);
+  return s + s * y;
+}
```
@Prabhuk I also would like to have your input on static storage inside function body, does it cause the same bloat it did with symbols before? or is it fine to leave them there?

lntue replied:
constants inside function body don't need `static`, just `constexpr`.

bassiounix commented on libc/src/mathvec/generic/expf.cpp:
```diff
@@ -0,0 +1,16 @@
+
+namespace LIBC_NAMESPACE_DECL {
+
+cpp::simd<float> expf(cpp::simd<float> x) { return mathvec::expf(x); }
```
Shouldn't this be wrapped with `LLVM_LIBC_FUNCTION`?
I'm not aware of the dynamics of this new library but from my understanding every public entry/symbol in libc should be wrapped with this macro.

```suggestion
LLVM_LIBC_FUNCTION(cpp::simd<float>, expf, (cpp::simd<float> x)) {
  return mathvec::expf(x);
}
```

So my question becomes: is this public symbol in `libmvec`?
cc @michaelrj-google @lntue @jhuber6

bassiounix commented on libc/src/__support/mathvec/expf.h:
```diff
@@ -0,0 +1,84 @@
+#include "src/__support/CPP/simd.h"
+#include "src/__support/FPUtil/FPBits.h"
+#include "src/__support/common.h"
+#include "src/__support/mathvec/expf_utils.h"
```
```suggestion
#include "expf_utils.h"
#include "src/__support/CPP/simd.h"
#include "src/__support/FPUtil/FPBits.h"
#include "src/__support/common.h"
```

bassiounix reviewed (changes requested):
Just some nits and removal of `static` qualifier inside header files.

bassiounix commented:
> I've said this before

Sorry I didn't catch that, my bad :)

> These need to be private to the TU as they encode ABI information

Can't we achieve this by marking the symbol visibility as private instead of static?

DylanFleming-arm committed:
Cleaned up CMake files and removed unneeded statics

DylanFleming-arm committed:
Fixed ABI errors

bassiounix commented on libc/src/__support/mathvec/expf.h:
```diff
@@ -0,0 +1,84 @@
+namespace mathvec {
+
+template <size_t N>
+LIBC_INLINE cpp::simd<double, N> static inline_exp(cpp::simd<double, N> x) {
```
```suggestion
LIBC_INLINE static cpp::simd<double, N> inline_exp(cpp::simd<double, N> x) {
```

bassiounix reviewed (approved):
Just some suggestions to maintain consistency in library code. But it's fine either way.

@Prabhuk we might need to disable libmvec on baremetal if it causes any issues because of `static`, we can't remove it from here (from the discussion above).

Other than that, LGTM.

DylanFleming-arm committed:
Reordered function qualifiers
